src/System.Net.Http/src/System/Net/Http/Managed/HttpSystemProxy.cs

```diff
@@ -0,0 +1,162 @@
+                    // Proxy failures are currently ignored by managed handler.
+                    return null;
+                }
+            }
+            return new HttpSystemProxy(proxyHelper, sessionHandle);
+        }
+
+        private HttpSystemProxy(WinInetProxyHelper proxyHelper, SafeWinHttpHandle sessionHandle)
+        {
+            _proxyHelper = proxyHelper;
+            _sessionHandle = sessionHandle;
+
+            if (proxyHelper.ManualSettingsOnly)
+            {
+                _proxyUri = GetUriFromString(proxyHelper.Proxy);
+
+                if (!string.IsNullOrWhiteSpace(proxyHelper.ProxyBypass))
+                {
+                    // Process bypass list for manual setting.
+                    string[] list = proxyHelper.ProxyBypass.Split(';');
+                    List<string> tmpList = new List<string>();
+
+                    foreach (string value in list)
+                    {
+                        string tmp = value.Trim();
+                        if (tmp == "<local>")
+                        {
+                            _bypassLocal = true;
+                            continue;
+                        }
+                        if (tmp.Length > 0)
+                        {
+                            tmpList.Add(tmp);
+                        }
+                    }
+                    if (tmpList.Count > 0)
+                    {
+                        _bypass = tmpList.ToArray();
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// This function will evaluate given string and it will try to convert
+        /// it to Uri object. The string could contain URI fragment, IP address and  port
+        /// tuple or just IP address or name. It will return null if parsing fails.
+        /// </summary>
+        private static Uri GetUriFromString(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                return null;
+            }
+
+            if (!value.Contains("://"))
+            {
+                value = "http://" + value;
+            }
+
+            if (Uri.TryCreate(value, UriKind.Absolute, out Uri uri) &&
+                    (uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps))
+            {
+                // We only support http and https for now.
+                return uri;
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Gets the proxy URI. (iWebProxy interface)
+        /// </summary>
+        public Uri GetProxy(Uri uri)
+        {
+            if (_proxyHelper.ManualSettingsOnly)
+            {
+                return _proxyUri;
+            }
+            var proxyInfo = new Interop.WinHttp.WINHTTP_PROXY_INFO();
+            if (_proxyHelper.GetProxyForUrl(_sessionHandle, uri, out proxyInfo))
+            {
+                return GetUriFromString(Marshal.PtrToStringUni(proxyInfo.Proxy));
+            }
```

Comment:
It looks like there's a leak here.  I believe you need to Marshal.FreeHGlobal members of proxyInfo like Proxy when you're done with the struct.  From https://msdn.microsoft.com/en-us/library/windows/desktop/aa384097(v=vs.85).aspx:
```
pProxyInfo [out]
A pointer to a WINHTTP_PROXY_INFO structure that receives the proxy setting. This structure is then applied to the request handle using the WINHTTP_OPTION_PROXY option. Free the lpszProxy and lpszProxyBypass strings contained in this structure (if they are non-NULL) using the GlobalFree function.
```
And I see WinHttpHandler doing that here:
https://github.com/dotnet/corefx/blob/f345a2251bcb57d3a64e03efc63dafb76319dc20/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L1023-L1027

Reply:
Would it be possible to allocate needed buffer on stack somehow? 
This does not need to live long or outside of this function.

Is there any good way how to detect memory leaks in out test infrastructure?

Reply:
>Is there any good way how to detect memory leaks in out test infrastructure?

I don't think there is especially for small leaks like this.